src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs

```diff
@@ -48,8 +46,29 @@ public async Task<string> GetSecretValue(Uri secretUri, CancellationToken cancel
 
             if (client != null)
             {
-                KeyVaultSecret secret = await client.GetSecretAsync(secretName, secretVersion, cancellationToken).ConfigureAwait(false);
-                secretValue = secret?.Value;
+                KeyVaultSecret secret;
+
+                // Try to load secret value from the cache first
+                secretValue = GetCachedSecretValue(key);
+
+                if (secretValue == null)
+                {
+                    // We dont have a cached secret value for this key vault reference.
+                    // Get the secret from Key Vault and update the cache.
+                    secret = await client.GetSecretAsync(secretName, secretVersion, cancellationToken).ConfigureAwait(false);
+                    secretValue = secret?.Value;
+
+                    if (secret != null)
+                    {
+                        UpdateCachedKeyVaultSecrets(key, secretValue);
+                    }
+                    else
+                    {
+                        // Secret may have been deleted from KeyVault.
+                        // Delete the secret from cache too.
+                        RemoveExpiredSecretFromCache(key);
```

Comment:
NIT: just for readability, the secret deleted from keyvault may or may not expired, so I would avoid using "ExpiredSecret" in the function name.